inject/inject/src/main/java/io/helidon/inject/ServicePlanBinder.java

```diff
@@ -0,0 +1,153 @@
+                                                                      toManagers(descriptors));
+
+        injectionPlan.put(injectionPoint, new IpPlan<>(supply, descriptors));
+        return this;
+    }
+
+    @Override
+    public ServiceInjectionPlanBinder.Binder bindList(Ip injectionPoint, ServiceInfo... descriptors) {
+        ServiceSupplyList<?> supply = new ServiceSupplyList<>(Lookup.create(injectionPoint),
+                                                              toManagers(descriptors));
+
+        injectionPlan.put(injectionPoint, new IpPlan<>(supply, descriptors));
+        return this;
+    }
+
+    @Override
+    public ServiceInjectionPlanBinder.Binder bindSupplier(Ip injectionPoint, ServiceInfo descriptor) {
+        ServiceSupply<?> supply = new ServiceSupply<>(Lookup.create(injectionPoint),
+                                                      toManagers(descriptor));
+
+        injectionPlan.put(injectionPoint, new IpPlan<>(() -> supply, descriptor));
+        return this;
+    }
+
+    @Override
+    public ServiceInjectionPlanBinder.Binder bindSupplierOfOptional(Ip injectionPoint, ServiceInfo... descriptors) {
+        ServiceSupplyOptional<?> supply = new ServiceSupplyOptional<>(Lookup.create(injectionPoint),
+                                                                      toManagers(descriptors));
+
+        injectionPlan.put(injectionPoint, new IpPlan<>(() -> supply, descriptors));
+        return this;
+    }
+
+    @Override
+    public ServiceInjectionPlanBinder.Binder bindSupplierOfList(Ip injectionPoint, ServiceInfo... descriptors) {
+        ServiceSupplyList<?> supply = new ServiceSupplyList<>(Lookup.create(injectionPoint),
+                                                              toManagers(descriptors));
+
+        injectionPlan.put(injectionPoint, new IpPlan<>(() -> supply, descriptors));
+        return this;
+    }
+
+    @Override
+    public ServiceInjectionPlanBinder.Binder bindOptionalOfSupplier(Ip injectionPoint, ServiceInfo... descriptors) {
+        // we must resolve this right now, so we just use the first descriptor, and hope the user did not inject
+        // this in a wrong scope
+        ServiceSupply<?> supply = new ServiceSupply<>(Lookup.create(injectionPoint),
+                                                      toManagers(descriptors[0]));
+        injectionPlan.put(injectionPoint, new IpPlan<>(() -> Optional.of(supply), descriptors));
+        return this;
+    }
+
+    @Override
+    public ServiceInjectionPlanBinder.Binder bindListOfSuppliers(Ip injectionPoint, ServiceInfo... descriptors) {
+        Lookup lookup = Lookup.create(injectionPoint);
+        // we must resolve the list right now (one for each descriptor)
+        List<ServiceSupply<Object>> supplies = Stream.of(descriptors)
+                .map(this::toManagers)
+                .map(it -> new ServiceSupply<>(lookup, it))
+                .toList();
```

Comment:
toList!

Reply:
I will do profiling once we have a web application working (as it will use injection a lot, and request scope). That should reveal the points that require attention.